ConfirmAddSuggestToken: replace mapDispatchToProps & mapStateToProps; Add Tests

digiwand opened the pull request:
**Fixes:** #

No issue

**Explanation:**  

Update ConfirmAddSuggestToken. No logic changes. We recently converted this component to a functional component. This work is a follow-up work for https://github.com/MetaMask/metamask-extension/pull/13377 as mentioned here https://github.com/MetaMask/metamask-extension/pull/13402#issuecomment-1026430441

- replace mapDispatchToProps with useDispatch
- replace mapStateToProps with useSelector
-  [In Progress] write tests

**Manual testing steps (For EtherScan. Can replicate on another platform, e.g., CoinGecko):**

- View an ERC20 token on EtherScan ([example](https://etherscan.io/token/0xdac17f958d2ee523a2206206994597c13d831ec7))
- Click on the "More" dropdown in "Profile Summary"
- Click on "Add Token to MetaMask (Web3)"
- Add token using MetaMask
- Repeat step 3
- Observe the "already listed" warning message. This should appear when the user is attempting to re-add a token with a token address already in their wallet.
- Test "cancel" button closes MetaMask

digiwand committed:
ConfirmAddSuggestedToken: component.js -> .js

digiwand committed:
replace mapStateToProps w/ useSelector
ConfirmAddSuggestedToken

digiwand committed:
replace mapDispatchToProps w/ useDispatch
ConfirmAddSuggestedToken

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand committed:
ConfirmAddSuggestedToken: useHistory

digiwand committed:
ConfirmAddSuggestedToken: revert scss
reuse ConfirmImportToken styles for now

digiwand committed:
ConfirmAddSuggestToken: add tests

digiwand committed:
ConfirmAddSuggestedToken: rm unused controls

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand committed:
ConfirmAddSuggestedToken: rm snapshot

digiwand committed:
use `metamark.svg` instead of `metamask.svg`

adonesky1 commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:
```diff
@@ -51,18 +54,19 @@ function hasDuplicateSymbolAndDiffAddress(suggestedAssets, tokens) {
+    getMostRecentOverviewPage(state),
+  );
+  const suggestedAssets = useSelector(
+    (state) => state.metamask.suggestedAssets,
```
you could make a new selector for this:
in `ui/selectors/selectors.js`
```
export function getSuggestedAssets(state) {
  return state.metamask.suggestedAssets;
}

```

adonesky1 commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:
```diff
@@ -51,18 +54,19 @@ function hasDuplicateSymbolAndDiffAddress(suggestedAssets, tokens) {
+  const dispatch = useDispatch();
+  const history = useHistory();
+
+  const mostRecentOverviewPage = useSelector((state) =>
+    getMostRecentOverviewPage(state),
+  );
```
This can be simplified to this:
```suggestion
  const mostRecentOverviewPage = useSelector(getMostRecentOverviewPage);
  );
```

adonesky1 commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:
```diff
@@ -51,18 +54,19 @@ function hasDuplicateSymbolAndDiffAddress(suggestedAssets, tokens) {
+  const suggestedAssets = useSelector(
+    (state) => state.metamask.suggestedAssets,
+  );
+  const tokens = useSelector((state) => state.metamask.tokens);
```
A selector for this already exists in 
`ui/ducks/metamask/metamask.js`

```suggestion
  const tokens = useSelector(getTokens);
```

digiwand committed:
ConfirmAddSuggestedToken: update useSelectors

digiwand committed:
Revert "ConfirmAddSuggestedToken: revert scss"

This reverts commit 07aed4576caaf247cb38ad481d79a339ba3d3947.

digiwand commented:
@adonesky1 

Thank you for the review! This makes sense. I applied the suggestions to the code with the import statements. 

re: https://github.com/MetaMask/metamask-extension/pull/13402#issuecomment-1026430441 
I thought more about the single selector in the newest `confirm-add-suggested-token/index.scss` file. We should not move the style into `confirm-import-token/index.scss` since the style does not exist for the `ConfirmImportToken`. Then I considered moving the `ConfirmImportToken` styles into this `.scss` file. Even though I'm usually okay duplicating or even triplicating code before making it reusable code, I'm afraid that copying the styles will require more maintenance and can more easily lead us to create style disparities. With these thoughts, what do you think about keeping the styles fragmented as they exist now? Or, is there another idea you had in mind to consolidate the styles?

digiwand committed:
ConfirmAddSuggestedToken:  === undefined
w/ strict mode

digiwand committed:
ConfirmAddSuggestedToken: use useNewMetricEvent

digiwand committed:
ConfirmAddSuggestedToken: run useEffect once

digiwand committed:
ConfirmAddSuggestedToken: rm unused async

digiwand committed:
ConfirmAddSuggestedToken: rm mock redux store

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

adonesky1 commented:
> Then I considered moving the ConfirmImportToken styles into this .scss file. Even though I'm usually okay duplicating or even triplicating code before making it reusable code, I'm afraid that copying the styles will require more maintenance and can more easily lead us to create style disparities

I think my preference for the short term would be to duplicate these styles in `confirm-add-suggested-token/index.scss`. Since these are considered different components for now they should have there own styles. We're duplicating the `react` so it feels messy to me to not duplicate the styles. If we determine that these components are almost identical minus a few conditional changes, we ought to consolidate them into one react component with some conditional rendering to vary across use cases. This is just my opinion 😄

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand committed:
ConfirmAddSuggestedToken: duplicate ConfirmImportToken styles

digiwand commented:
@adonesky1  Thank you for explaining your opinion here because I was close to being indifferent about duplicating the styles and reusing the styles. I went with duplicating the styles! I also went with not consolidating the two components and styles with a new react component. I can see a benefit in creating a page abstraction if we reuse this page UI in the future. 

What I think could be helpful right now, and in a different PR, would be to consolidate the token or asset lists. Thing is, there seem to be a few variations of lists and list items in our repo. We currently have an AssetList and TokenList Component/UI that is not used on these pages. I can keep this in mind for the future as I learn more about this codebase.

digiwand commented:
It might be worth mentioning here that by reading the code, it would appear we could expect the app to redirect to the overview page, `history.push(mostRecentOverviewPage);`, after clicking "Add Tokens". Instead, through CoinGecko or EtherScan, the popup closes after clicking "Add Tokens". I suspect this may be expected behavior with the way CoinGecko and EtherScan use the MM API. 

Again, no UI changes have been introduced in this PR. These GIFs below show that the previous and newer "Add Tokens" click remains the same.

previous MM UI w/ "Add Tokens" click:
<img src="https://user-images.githubusercontent.com/20778143/155352181-21b099f7-eaf1-453c-99c4-ac8700ceea96.gif" width="360" />

newer MM UI w/ "Add Tokens" click (same click behavior):
<img src="https://user-images.githubusercontent.com/20778143/155352688-a679c315-a7a0-4750-927b-d64eef516cfb.gif" width="360" />

danjm commented:
> It might be worth mentioning here that by reading the code, it would appear we could expect the app to redirect to the overview page, history.push(mostRecentOverviewPage);, after clicking "Add Tokens".

This is what happens if you have the add tokens confirmation open in the full screen tab or in the extension "popup" view (by clicking the fox icon in the top right of your browser). When interacting via the notification window, it should close upon confirmation as your screen capture displays

digiwand commented:
@danjm Thanks for clarifying this! I'd also like to clarify that the MM is opened in a popup window and not the popup extension in the GIFs above. What Dan mentions above is in line with expected behaviors.

danjm commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:
```diff
@@ -117,11 +118,18 @@ const ConfirmAddSuggestedToken = (props) => {
   }, [suggestedAssets, tokens, t]);
 
+  useEffect(() => {
+    if (Object.keys(addedToken).length) {
```
This sustains the behaviour of tracking the metrics event when the submit button is clicked, but it does so indirectly. Previously, the metrics event would be tracked as part of the onclick method call, but not it is tracked as a result of a state change that is created by the onclick method call.

This could cause confusion for a reader of the code, who might expect `useEffect` to be responding to a state change that is outside of the control of this component. Also, it increases the number of places in which a bug could be introduced in the future. Previously, the path was `onclick call -> metric event method call`, now the path is `onclick call -> state update -> useEffect recognition of state update -> metric event method call`. So there are more potential steps that could be accidentally broken in the future. Finally, as a general rule, if useEffect can be avoided for something simpler, that generally make components more readable and maintainable.

And in this case, there is some risk that this might not always work as desired when there are multiple `suggestedAssets` 

I would suggest that we continue just to call the metrics event method within the onclick handler. To do this, you could put the whole onclick handler function in a function defined in the body of the component function and then pass that to `onClick`.

I think this would require using the existing `tokenAddedEvent` method

digiwand replied:
@danjm 

Thank you for calling this out! As mentioned in this past Wednesday's standup, I have updated the logic to call `trackEvent` instead of `useNewMetricEvent` directly. 

I agree using `useEffect` can cause unwanted side-effects as `suggestedAssets` is susceptible to change. 

I did go with calling the click handler from the `onClick` function. I think the biggest hurdle was finding a way to use `useNewMetricEvent`, and the key was not to use this when we have dynamic params.  Since `useNewMetricEvent` returns a custom hook, we cannot call it inside another callback which would cause some limitations.

danjm commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:
```diff
@@ -117,11 +118,18 @@ const ConfirmAddSuggestedToken = (props) => {
       history.push(mostRecentOverviewPage);
     }
-  }, [history, suggestedAssets, mostRecentOverviewPage]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Why was `// eslint-disable-next-line react-hooks/exhaustive-deps` needed?

digiwand replied:
Passing an empty array will ensure this hook is only called once. This is the closest I think we could come to replicating the `componentDidMount` which is what was used when this was a class component. 

see "Note:" at the bottom of ref: https://reactjs.org/docs/hooks-effect.html

If we were to pass `suggestedAssets` as a dependent property here, `history.push(mostRecentOverviewPage);` could override other redirects that happen after altering the `suggestedAssets`. For instance, in the `ConfirmImportToken` page, we redirect the user to the token page after the token has been added. If we were to replicate that behavior here, using `useEffect` with `[suggestedAssets]` dependency list would cause a race condition.

It is recommended that we keep the ESLint warning for `react-hooks/exhaustive-deps` and not use this rule as a last resort.

ref: https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often
> Only do this if you couldn’t find a better alternative, as relying on mutation makes components less predictable.

danjm replied:
Okay, I think you can do something like this:

```
const goBackIfNoSuggestedAssetsOnFirstRender = () => {
    if (!suggestedAssets.length) {
      history.push(mostRecentOverviewPage);
    }
}

useEffect(() => {
    goBackIfNoSuggestedAssetsOnFirstRender();
}, []);
```

digiwand replied:
lgtm! Thanks for the function name suggestion. Updated the repo with this suggestion

digiwand replied:
@danjm 

lgtm! Thanks for the function name suggestion. Updated the repo with this suggestion

danjm replied:
I think `// eslint-disable-next-line react-hooks/exhaustive-deps` can be removed now?

digiwand replied:
@danjm @adonesky1 

Removing `// eslint-disable-next-line react-hooks/exhaustive-deps` will enable the warning message again on `[]`.  Adding `goBackIfNoSuggestedAssetsOnFirstRender` as a dependency would cause useEffect to rerun again on each render unless we wrapped it in useCallback, or declare the function inside the useEffect hook. With either of these options, we would run into the need for the empty dependency list, `[]`, either on the useCallback hook or the useEffect hook. 

Is there another way we could go about removing `// eslint-disable-next-line react-hooks/exhaustive-deps` and only rendering this check once on initial load? 

ESLint warning after replacing `[]` with `[goBackIfNoSuggestedAssetsOnFirstRender]` in useEffect:
<img width="1065" alt="Screenshot 2022-03-03 at 1 20 56 PM" src="https://user-images.githubusercontent.com/20778143/156606417-f7896187-aee5-466c-9a59-2dd9bc63e267.png">

adonesky1 replied:
Does it not work to wrap `goBackIfNoSuggestedAssetsOnFirstRender` in `useCallback` add all dependencies to the dependency array and then add `goBackIfNoSuggestedAssetsOnFirstRender` to the dependency array of the `useEffect`?

digiwand replied:
I don't think so. The initial logic lives in `componentDidMount`, so the thing is, we only want to run this logic on initial load or initial render in this case. I think with this suggestion, if the list of suggestions changes, for instance on `acceptWatchAsset`, then this `useEffect` will run because one of the dependencies of `useCallback` has changed. This causes a race condition between `history.push` here, and the one in the `Add Tokens` click handler.

digiwand replied:
@adonesky1 

I don't think so. The initial logic lives in componentDidMount, so the thing is, we only want to run this logic on initial load or initial render in this case. I think with this suggestion, if the list of suggestions changes, for instance on acceptWatchAsset, then this useEffect will run because one of the dependencies of useCallback has changed. This causes a race condition between history.push here, and the one in the Add Tokens click handler.

jpuri replied:
I can think of an edge case here, what is `suggestedAssets` are not loaded on first render ?

digiwand replied:
@jpuri 

This is a good question. 

What I see happening is when the `wallet_watchAsset` [API request](https://docs.metamask.io/guide/registering-your-token.html) is requested, the `watchAssetHandler` RPC middleware calls [TokensController#watchAsset](https://github.com/MetaMask/controllers/blob/main/src/assets/TokensController.ts#L393). This method validates the asset, updates the redux `suggestedAssets` state with the new asset, and then emits `pendingSuggestedAsset`. `pendingSuggestedAsset` will then call [background.js#openPopup](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/background.js#L605). Unless I am missing something, I think we can safely assume that it will be there before the popup is opened. 

I also think it is safe to make this change due to the order that the hooks run.  `componentDidMount` runs after mounting and before painting. `useEffect` runs after the first paint. We had replaced `componentDidMount` in https://github.com/MetaMask/metamask-extension/pull/13377/files, so the logic is near the same here.

danjm reviewed (changes requested):
Left one suggestion and one question

digiwand committed:
ConfirmAddSuggestedToken: update metrics logic

brad-decker commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:
```diff
@@ -117,11 +106,40 @@ const ConfirmAddSuggestedToken = (props) => {
+      suggestedAssets.map(async ({ asset, id }) => {
+        await dispatch(acceptWatchAsset(id));
+
+        metricsEvent({
```
[REQ] change this to `trackEvent` which is the name of the context function for the new context, but also the useNewMetricsEvent hook calls it metricsEvent which is wrong so i get why there is like 5 layers of misdirection. 

Maybe i was trying to encode job security here? 🤷🏻

digiwand replied:
updated, thank you! @brad-decker

digiwand committed:
ConfirmAddSuggestedToken: metricEvent-> trackEvent

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

danjm reviewed (dismissed):
Looking good! One needed change left that I have commented on

digiwand committed:
ConfirmAddSuggestedToken:clean useEffect w/ new fn
- PR suggestion: https://github.com/MetaMask/metamask-extension/pull/13526#discussion_r818007393

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

jpuri commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.test.js:
```diff
@@ -0,0 +1,142 @@
+import React from 'react';
```
Good work on test coverage 👍

jpuri commented on ui/pages/confirm-add-suggested-token/confirm-add-suggested-token.js:
```diff
@@ -1,16 +1,21 @@
 function getTokenName(name, symbol) {
-  return typeof name === 'undefined' ? symbol : `${name} (${symbol})`;
+  return name === undefined ? symbol : `${name} (${symbol})`;
 }
```
👍

digiwand replied:
per @mcmire's suggestion here https://github.com/MetaMask/metamask-extension/pull/13594#discussion_r805027489 🙂 ✨

adonesky1 reviewed (dismissed):
Looks like you need a rebase 😢  but otherwise LGTM!

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand commented:
@jpuri @adonesky1 Thank you! I rebased and verified this feature works.

darkwing commented on ui/pages/confirm-add-suggested-token/index.scss:
```diff
@@ -1,9 +1,77 @@
+        @include H1;
+
+        color: var(--scorpion);
+        margin-right: 8px;
```
`margin-inline-end` would be good for RTL.

digiwand replied:
It turns out that the `.token-balance` class is not summoned here. `TokenBalance` component exists in this component. I removed these styles. We should be removing `.token-balance` from `ui/pages/confirm-import-token/index.scss` in a separate PR.

digiwand replied:
Update: spoke with @darkwing  about this and we don't think `margin-inline-end` is necessary here. We are creating a follow-up task to update the spacing around this element after some margin/padding concerns we noticed.

darkwing commented on ui/pages/confirm-add-suggested-token/index.scss:
```diff
@@ -1,9 +1,77 @@
+  }
+
+  &__token-icon {
+    margin-right: 12px;
```
`margin-inline-end` would be good for RTL.

digiwand replied:
@darkwing  
`margin-inline-end` is new to me. I see the benefit when the direction changes. How does it help here?

digiwand replied:
Update: spoke with @darkwing about this and we don't think margin-inline-end is necessary here

darkwing reviewed (changes requested):
This looks awesome!  Firefox was spotty but I noticed it's spotty on `develop` branch.  The popup usually appears on the first click, but doesn't appear to ever do so again.  I don't think it's related to your changes.

digiwand commented:
@darkwing 

Thanks for mentioning this! I am confirming your observation and that this does not appear to be related to any of these changes.

I also observed this page is spotty in Firefox for this branch and `develop` while it works fine in Brave, Chrome, and Opera. Then, I checked out the last commit before https://github.com/MetaMask/metamask-extension/pull/13377 was merged, d25f0e5817ecdc8e37bfb144e4b7ebf299233421. This was the commit before I made any refactoring changes. I verified this page also works in Opera, Chrome, and Brave. Firefox behaves the same for this commit as it does on this branch and on `develop`. 

The behavior that we are observing in Firefox is that after adding the token once, the popup will not open again. On CoinGecko, after adding the token once via MM, each time the user clicks "Add to MetaMask" again, the "Add Token" notification toast will appear, and the popup will not work again.

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand committed:
ConfirmAddSuggestedToken: rm unused .token-balance styles

darkwing reviewed (approved):
This is awesome!  Well done!

digiwand commented:
Thanks for the patience and thoughtful reviews everyone!

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand committed:
Merge branch 'develop' into clean/confirm-to-fc/confirm-add-suggested-token-pt-2

digiwand commented:
@darkwing  

Found something. Looks like this issue, opened in June 2021, may be related to the finicky behavior we were observing in Firefox here: 
https://github.com/MetaMask/metamask-extension/issues/11377
